rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,473 @@
+     SE_StreamExecutor* (*create_stream_executor)(TF_Status* status),
+     void (*delete_device)(SE_Device* device),
+     void (*delete_stream_executor)(SE_StreamExecutor* stream_executor);
+);
+
+TF_CAPI_EXPORT extern void SE_RegisterPlatform(
+     char* name,
+     size_t name_len,
+     SE_Platform* platform,
+     TF_Status* status);
+
+#ifdef __cplusplus
+} /* end extern "C" */
+#endif
+```
+
+## PlatformId
+
+`SE_PlatformId.id` should be set to a unique identifier.
+
+## Usage example
+
+Define functions that create and delete `SE_Device` and `SE_StreamExecutor`:
+
+```cpp
+SE_Device* create_device(SE_Options* options, TF_Status* status) {
+  SE_Device* se = new SE_Device();
+  se->device_handle = get_my_device_handle();
+  ...
+  return se;
+}
+SE_StreamExecutor* create_stream_executor(TF_Status* status) {
+  SE_StreamExecutor* se_fns = new SE_StreamExecutor();
+  se->memcpy_from_host = my_device_memcpy_from_host_function;
+  ...
+  return se;
+}
+void delete_device(SE_Device* device) {
+  -- delete device handle here --
+  ...
+  delete device;
+}
+void delete_stream_executor(SE_StreamExecutor* stream_executor) {
+  delete stream_executor;
+}
+```
+
+Create a new platform using `SE_NewPlatform` and register it using
+`SE_RegisterPlatform`:
+
+```cpp
+void RegisterMyCustomPlatform() {
+  static plugin_id_value = 123;
+  SE_PlatformId id;
+  id.id = &plugin_id_value;
+  int visible_device_count = 2;
+
+  SE_Platform* custom_platform = SE_NewPlatform(
+     id, visible_device_count,
+     create_device, create_stream_executor,
+     delete_device, delete_stream_executor);
+
+  TF_Status* status = TF_NewStatus();
+  std::string name = "MyCustomDevice";
+  SE_RegisterPlatform(
+     name.c_str(), name.size(),
+     custom_platform,
+     status);
+}
+```
+
+Use static initialization to register the new platform:
+
+```cpp
+static bool IsMyCustomPlatformRegistered = []() {
+ RegisterMyCustomPlatform();
+ return true;
+}();
+```
+
+## Stream/Timer/Event representation
+
+API extension would require defining SE\_Stream\_st, SE\_Event\_st and
+SE\_Timer\_st structs. From the point of view of TensorFlow, we will treat their
+pointers as opaque.
+
+Underneath, StreamExecutor will rely on customized implementations of
+[StreamInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=114?q=TimerInterface&ss=tensorflow%2Ftensorflow),
+[TimerInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=145?q=TimerInterface&ss=tensorflow%2Ftensorflow)
+and
+[EventInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=76?q=TimerInterface&ss=tensorflow%2Ftensorflow).
+For example, Stream customization might look as follows:
+
+```cpp
+class CStream : public StreamInterface {
+ public:
+  explicit CStream(TF_StreamExecutor* device,
```

Comment:
does TF_StreamExecutor* device means SE_Device? is TF_StreamExecutor the same as SE_StreamExecutor?